src/coreclr/nativeaot/Runtime/eventpipe/ep-rt-aot.cpp

```diff
@@ -85,12 +84,66 @@ ep_rt_aot_sample_profiler_write_sampling_event_for_threads (
 {
 }
 
+#if defined (__linux__)
+extern char *__progname;
+#endif
+
 const ep_char8_t *
 ep_rt_aot_entrypoint_assembly_name_get_utf8 (void) 
 {
-    // shipping criteria: no EVENTPIPE-NATIVEAOT-TODO left in the codebase
-    // TODO: Implement EventPipe assembly name - return filename in nativeaot?
-    return reinterpret_cast<const ep_char8_t*>("");
+    // Although PalGetModuleFileName exists, non-windows implementation has issues and we use __progname for Linux.
+    // Cannot use __cpp_threadsafe_static_init feature since it will bring in the C++ runtime and need to use threadsafe way to initialize entrypoint_assembly_name
+    static const ep_char8_t * entrypoint_assembly_name = nullptr;
+    if (entrypoint_assembly_name == nullptr) {
+        const ep_char8_t * entrypoint_assembly_name_local;
+#ifdef HOST_WINDOWS
+        const TCHAR * wszModuleFileName = NULL;
+        if(PalGetModuleFileName(&wszModuleFileName, nullptr) == 0)
+            entrypoint_assembly_name_local = reinterpret_cast<const ep_char8_t*>("");
+        else {
+            const wchar_t* process_name = wcsrchr(wszModuleFileName, '\\');
+            if (process_name != NULL) {
+                const wchar_t* extension = wcschr(process_name + 1, '.');
+                if (extension != NULL) {
+                    // We don't want to include the first '\'
+                    size_t len = extension - (process_name + 1);
+                    wchar_t* process_name_wo_ext = reinterpret_cast<wchar_t *>(malloc(len + 1));
+                    wcsncpy(process_name_wo_ext, process_name + 1, len);
+                    process_name_wo_ext[len] = L'\0';
+                    const ep_char16_t* process_name_wo_ext_l = reinterpret_cast<const ep_char16_t *>(process_name_wo_ext);
+                    entrypoint_assembly_name_local = ep_rt_utf16_to_utf8_string(process_name_wo_ext_l, -1);
+                } else {
+                    const ep_char16_t* process_name_l = reinterpret_cast<const ep_char16_t *>(process_name + 1);
+                    entrypoint_assembly_name_local = ep_rt_utf16_to_utf8_string(process_name_l, -1);
+                }
+            }
+            else
+                entrypoint_assembly_name_local = reinterpret_cast<const ep_char8_t*>("");
+        }
+#else
+#if defined (__linux__)
+        size_t program_len = strlen(__progname);
+        ep_char8_t *process_name =  reinterpret_cast<ep_char8_t *>(malloc(program_len + 1));
+        if (process_name == NULL)
+            entrypoint_assembly_name_local = reinterpret_cast<const ep_char8_t*>("");
+        else {
+            memcpy (process_name, __progname, program_len);
+            process_name[program_len] = '\0';
+            char *dot = strrchr(process_name, '.');
```

Comment:
I think it'd make more sense to check for the extension before allocating and copying into `process_name`. That way you can just allocate / copy as much as is needed.